### PR TITLE
maximize back compat

### PR DESCRIPTION
**Pull request by ljharb**
This PR switches from jest to tape, so that we aren't limited by our test runner's engine support.

It also switches dequal (which requires node 6+) to deep-equal-json, which has minimal dependencies and so should still meet the constraints that motivated switching to dequal in the first place.

It also expands the test matrix so that all supported engines are tested, now and moving forward.

### CODE REVIEW

File: .github/workflows/test.yml
Comment (benmccann) [inline .github/workflows/test.yml:24]: what does this do that's different from the standard `actions/setup-node`?

**Review by benmccann — CHANGES_REQUESTED** [review]
> It also switches dequal (which requires node 6+) to deep-equal-json, which has minimal dependencies and so should still meet the constraints that motivated switching to dequal in the first place.

I really don't think that's accurate. This new library introduces 16 additional dependencies: https://npmgraph.js.org/?q=deep-equal-json

Perhaps we can find a library with no additional dependencies that fits your requirements for Node 4 support?

**Comment by ljharb** [comment]
Indeed, it doesn't have as few dependencies, but it has fewer than deep-equal which was used previously. If it can be done with fewer dependencies and the same reliability and compat, then I've love to review some PRs on deep-equal-json (and deep-equal).

This PR extends support down to node _0.4_ and equivalent browsers, which is useful because accessibility means including *everyone*, even those few users on old browsers.

File: .github/workflows/test.yml
Comment (ljharb) [reply .github/workflows/test.yml:25]: It supports a wider range of node versions, and handles a lot of edge cases for npm installing in older ones. It also uses nvm to install node, and `nvm install-latest-npm`, which is the only way I'm aware of to reliably get the latest possible working version of npm on older nodes.

**Comment by benmccann** [comment]
Can you show me where `dequal` would not work on an old browser?

**Comment by ljharb** [comment]
Fair, I hadn't looked at its code, just its engines declaration (which is important).

The "reliability" part includes a lot of things that https://github.com/lukeed/dequal/blob/master/src/lite.js doesn't cover - `delete Date.prototype.getTime` after the module is imported but before it runs would crash it, eg, and the `.constructor` property is forgeable (can be faked, or deleted) and doesn't work across realms (iframes, node's vm module, etc).

**Comment by benmccann** [comment]
> delete Date.prototype.getTime

Frankly, if you do that, you get what you deserve. I'd consider that breakage a feature rather than a bug as people should be prevented from engaging in such nonsense.

**Comment by Rich-Harris** [comment]
> Some initial thoughts: the "reliability" part includes a lot of things that https://github.com/lukeed/dequal/blob/master/src/lite.js doesn't cover

As we established previously, this simply isn't relevant here: https://github.com/A11yance/aria-query/pull/497#issuecomment-1579432306

As a user of this library I am strongly opposed to replacing `dequal` with a mess of dependencies that will bloat everyone's `node_modules` for absolutely no good reason whatsoever 

image

**Comment by ljharb** [comment]
It's not always just "you" - extensions, user code, etc all can be running in an application. Also, the person who suffers isn't the person who wrote that code - it's the end user who suffers, and they did nothing wrong.

For what it's worth I hear you about wanting to minimize install/bundle size (I assume that's the metric that matters, since i can't figure out how dependency count would matter beyond those things), but I think what I'm not understanding is how correctness, robustness, reliability, and accessibility should all take a backseat to that constraint.

**Comment by ljharb** [comment]
@Rich-Harris if @lukeed is willing to expand the engines.node declaration (and testing matrix) to cover what's needed here, then that'd be fine. Since deep-equal-json already exists, this seemed more expedient.

**Comment by Rich-Harris** [comment]
> what's needed here

No-one needs Node 4 support. This is utterly absurd.

**Comment by ljharb** [comment]
I understand you two feel it's absurd - you're not alone in feeling that way. That doesn't mean it's not a requirement.

**Comment by benmccann** [comment]
> It's not always just "you" - extensions, user code, etc all can be running in an application. Also, the person who suffers isn't the person who wrote that code - it's the end user who suffers, and they did nothing wrong.

An extension that does that will immediately break everything. It will thusly have near zero users and does not need to be a consideration

> For what it's worth I hear you about wanting to minimize install/bundle size (I assume that's the metric that matters, since i can't figure out how dependency count would matter beyond those things)

Dependency count actually matters greatly to us. https://learn.svelte.dev/ installs this version into a web container in the browser. 16 extra dependencies is an extra 16 HTTP calls.

> I understand you two feel it's absurd - you're not alone in feeling that way. That doesn't mean it's not a requirement. (do note that node 4 still has many millions of downloads according to node's own access logs)

Engineering is about trade-offs. Version 3.2.0 was released without Node 4 support over a year ago. Not a single person has filed an issue during all that time.

Since 3.2.0 was released, a new major version of this library was released. If this change were going to be made it should at least be done as a backport to version 3 and we just update the changelog for version for to explicitly state that support for Node 4 has been dropped. But even that seems unnecessary to me

**Comment by lukeed** [comment]
Please stop

**Comment by ljharb** [comment]
@benmccann that might be an alternative path - to keep dequal, tighten up the engines.node on v4, and then do a backport to v3 that switches to an alternative dependency, since eslint-plugin-jsx-a11y isn't on v4 yet - your belief that nobody will care about node 4 may allow for the breaking change within v4 of restricting engines.node.

Engineering is indeed about tradeoffs, and I clearly think that dep count and install/bundle size are a worthy sacrifice in the face of compatibility, accessibility, correctness, and reliability.

@lukeed "Please stop" is not a constructive contribution to the discussion, but I'll take that to mean that you're not going to make any changes to dequal.

**Comment by ljharb** [comment]
> Dependency count actually matters greatly to us. https://learn.svelte.dev/ installs this version into a web container in the browser. 16 extra dependencies is an extra 16 HTTP calls.

it seems like it'd make a lot of sense then to have a webserver to load and cache all the code you need, so that the user's browser doesn't need to do the installs themselves - especially since many users, including in China, can't access npm directly. I assume that's not a new suggestion, though, and there's reasons you went with this approach.

**Comment by Rich-Harris** [comment]
> Engineering is indeed about tradeoffs, and I clearly think that dep count and install/bundle size are a worthy sacrifice in the face of compatibility, accessibility, correctness, and reliability.

Once again, since this point seems to have been missed: as far as this library is concerned, `deep-equal-json` has no advantages over `dequal` in this regard.

> it seems like it'd make a lot of sense then to have a webserver to load and cache all the code you need

FWIW we actually _do_ do this, the dependency is among others bundled into a zip file. But that's immaterial — in the common scenario, adding bloat like this _does_ affect users.

> Not a single person has filed an issue during all that time.

Which is the crux of the matter. Candidly, you are wasting everyone's time and bandwidth, not to mention this project's quota of actions minutes.

**Comment by KonnorRogers** [comment]
why is this needed?

 if people need support for older Node versions, they can add a polyfill to support the missing APIs.

**Comment by ljharb** [comment]
Public projects have no action minutes quotas.

@benmccann trying to summon a twitter mob isn't very professional or constructive, if I need to lock the thread then none of you will have any further chance to convince me of anything here.

**Comment by ljharb** [comment]
@Rich-Harris 
> FWIW we actually do do this, the dependency is among others bundled into a zip file. But that's immaterial — in the common scenario, adding bloat like this does affect users.

ok, but the stated use case and complaint was about HTTP requests on svelte's learn site - now it seems that it doesn't actually cause any additional HTTP requests for your users?

**Comment by Rich-Harris** [comment]
> it doesn't actually cause any additional HTTP requests for your users?

Of course it does — every time someone installs a package that depends on this one, including `svelte`. 

No-one is trying to summon a Twitter mob. It's just _incredibly_ frustrating to have to periodically deal with this nonsense. It's indistinguishable from trolling.

As a wiser man than me said,

> Please stop

**Comment by ljharb** [comment]
I've explained why it's not nonsense to me. Just because we disagree doesn't make either of us right - if you can't distinguish good faith differences in priorities from trolling then I'm not sure how I can help you, because it would be just as easy for me to call this pushback a delightful little brigading/trolling attempt.

I hear that you are frustrated. This isn't enjoyable for me either. But no, i will never stop doing what I think is best for the ecosystem, or for the vast quantities of users who I impact.

It seems like https://github.com/A11yance/axobject-query/pull/354#issuecomment-2183563329 is a plan that would meet my needs on v3, and continue to meet yours on v4, and then I could sign up to do the extra work of maintaining both release lines forever (on aria-query, too). Would that satisfy your constraints?

**Comment by Rich-Harris** [comment]
> Would that satisfy your constraints?

No it would not, because you're asking Luke to widen the test matrix to an extremely silly degree for _absolutely no good reason_. (Your predilections do not constitute a good reason.)

**Comment by ljharb** [comment]
Sorry if I wasn't clear. I mean that v4 would continue to use dequal, and v3 would use deep-equal-json.

(and your personal rubric doesn't define which reasons are "good" either, as I'm sure you're aware)

**Comment by Rich-Harris** [comment]
It's not my personal rubric, it's the unanimous view of everybody I've ever seen express an opinion on the matter.

> I mean that v4 would continue to use dequal, and v3 would use deep-equal-json

If that's what it takes to avoid having `deep-equal-json` in our dependency tree then so be it. I maintain that there is absolutely no point in adding it to v3, or it wouldn't have taken even you nine months to notice this, but your time is yours to waste as you please.

**Comment by 43081j** [comment]
This is ultimately to add support for older node versions from what I can gather. 

I'm guessing the lint plugin you have runs its tests against these versions and currently fails if you try update. But that is different from _consumers_ trying to use this in older versions. 

Are we sure there's actually consumer demand to support them?

**Comment by benmccann** [comment]
> It seems like https://github.com/A11yance/axobject-query/pull/354#issuecomment-2183563329 is a plan that would meet my needs on v3, and continue to meet yours on v4, and then I could sign up to do the extra work of maintaining both release lines forever (on aria-query, too). Would that satisfy your constraints?

Ultimately it's up to @jessebeach, but if that's the plan then why not just maintain your own fork of `axobject-query`? It seems like that would be equivalent while reducing the demands on her and noise in this repo

**Comment by ljharb** [comment]
@benmccann jesse has made me a maintainer of this library, so i'll be doing it here, and there won't be any additional demands on her time.

**Comment by ljharb** [comment]
@43081j that is a good question when considering making a major bump and dropping support! However, the dictates of semver mean that it's irrelevant if there's demand; one doesn't make breaking changes within a major line.

**Comment by ljharb** [comment]
(and for the record, it's not that i just recently noticed this issue, it's that my time is quite taken up by the demands of family, work, and shitloads of other unpaid open source work, so i wasn't able to get to it until now)

**Comment by benmccann** [comment]
> forks aren't helpful for the ecosystem, since it makes bundle sizes even larger by bypassing deduping

While I generally agree with this, I think we're in the same situation either way. Most of the ecosystem will be on v4 except for libraries that you maintain being indefinitely stuck on v3 and so those two different major versions won't be able to be deduped.

In any case, I'm glad we were able to come to a compromise that keeps minimal dependencies for most users.

**Comment by ljharb** [comment]
I've released v3.2.2 to unbreak the v3 line.

I'll shortly release a v4 patch that has reduced engine support, the `exports` field, and continues to use `dequal`, and I'll maintain both v3 and v4 together moving forwards.

**Comment by valadaptive** [comment]
> jesse has made me a maintainer of this library

I'd like to dig into this a bit more. As far as I can tell, you're not listed as a member of the @A11yance GH organization:
image

You're not in the `contributors` field [in package.json](https://github.com/A11yance/axobject-query/blob/3.x/package.json) either.

There is [no issue, open or closed, in this repository asking for help with maintenance or asking for somebody to transfer maintainership to](https://github.com/A11yance/axobject-query/issues?q=is%3Aissue+).

I am wondering:
- Is there any means by which somebody could tell that you are now a maintainer (possibly *the* only active maintainer now) of this project, prior to you making this change?
- How did this maintainership transfer come about? Who reached out to whom about becoming a maintainer? Did Jesse ask in some other channel for help maintaining this package, or did you reach out to her?

Given the recent `xz` supply-chain attacks, and [previous NPM supply chain attacks involving maintainership transfer](https://es-incident.github.io/paper.html), I find it somewhat concerning that maintainership of a package with 15 million downloads can be transferred without any sort of public announcement or even any prior indication that there was intent to do so.

Obviously, I don't believe you're a threat actor. But as a steward of the open-source community, I feel like you should be transparent in the leadership of the packages you maintain, and model the kind of behavior you would like to see in the ecosystem.

**Comment by HugeLetters** [comment]
I'm just some random guy who doesn't uses this package (but I use Svelte) and just stumbled upon this situation.
To me it feels extremely manipulative at best that these changes are being proposed under the guise of accessibility efforts while 
 - Jest was replaced with Tape. Tape is created by @ljharb 
 - dequal was replaced with deep-equal-json. deep-equal-json is created by @ljharb. deep-equal which was used before dequal was also created by @ljharb 
 - setup-node github action was replaced with some custom setup created by... @ljharb . No proper version lock, it's just `@main`, nice
 - Jest gh action was replaced with yet another custom setup created by @ljharb 
 - added mock-property package. created by @ljharb 
 - added object-inspect. created by @ljharb 
 - added object.values. created by @ljharb 

And these are only direct dependencies.

**Comment by ljharb** [comment]
@valadaptive you can see it on this very GitHub repo (where it says “collaborator”) and on the npm package page. GitHub org membership is never relevant.

@HugeLetters im not sure why you care what dev deps are used. note that i wasn’t involved in the original usage of deep-equal. Either way, i choose these tools because they’re the best ones to do the job that’s needed. That i maintain them means i can help ensure their fitness and stability, which makes them better choices.

As for “proper version lock”, it’d just be security theater to lock down a CI job that has no secrets and no possibility of collateral damage if compromised.

**Comment by valadaptive** [comment]
> @valadaptive you can see it on this very GitHub repo (where it says “collaborator”) and on the npm package page. GitHub org membership is never relevant.

I can see you're listed as a "Collaborator" on npm, but I can't find any reference to "collaborators" on GitHub, only "contributors" (which refers only to people who have had commits merged to `main` and has nothing to do with ownership). Does the "collaborator" status on npm appear when you *publish* a new version, or immediately after you're given publish rights?

Anyway, let's not get too sidetracked; 2 of my questions have still gone unaddressed. I'm still curious exactly how you became a maintainer of this repo, and whether you have commit rights to other projects that you have not yet exercised.

**Comment by ljharb** [comment]
It shows up immediately. As to your other questions, it was discussed publicly on eslint-plugin-jsx-a11y - but maintainership transfers are often discussed privately. I’m not sure why it’s any of your business where i have what privileges, and I’m not pleased with the implication behind your question.

**Comment by valadaptive** [comment]
> As to your other questions, it was discussed publicly on eslint-plugin-jsx-a11y - but maintainership transfers are often discussed privately. I’m not sure why it’s any of your business where i have what privileges, and I’m not pleased with the implication behind your question.

Would you mind linking to that discussion (assuming by "privately" you don't mean that it was discussed over email or somesuch)? I searched the issues and PRs for "maintainer" and "maintainership" and couldn't find anything, although I did find an instance of @jessebeach disagreeing with your philosophy of shipping tests in the final package.

I think it *is* people's business to know when the maintainership of one of their dependencies changes. Even when discussed privately, an announcement is typically posted in the issues or similar after the transfer has occurred.

I believe it is especially pertinent in this case because, as far as I can tell, your last interaction within this organization was last year when [you argued with the previous primary maintainer, @jessebeach, about removing one of your dependencies](https://github.com/A11yance/aria-query/pull/497).

I mean this genuinely--why do you keep volunteering for more and more unpaid work when you're already swamped with it 7 days a week? It seems genuinely exhausting foe you. Is it out of some sense of duty or obligation to certain members of the JS ecosystem?

**Comment by ljharb** [comment]
That’s very much not typical in my experience, but i see the appeal of such an announcement. I’d prefer not to link to the discussion because i don’t want to invite further back and forth elsewhere.

I appreciate your concern for me, but i don’t feel the need to launch into self-psychoanalysis with a stranger on the internet today.

**Comment by Rich-Harris** [comment]
In light of https://github.com/A11yance/axobject-query/pull/354#issuecomment-2184042343 I think it would be worthwhile to establish certain facts.

@ljharb do you have a financial interest, direct or otherwise, in usage of packages under your authorship? This isn't an implication or an accusation, but it's pertinent information for this discussion.

[Tidelift](https://tidelift.com/), for example, is a platform that pays open source maintainers that enter into an agreement, and the amount is [partially determined](https://tidelift.com/about/lifter) by usage:

> Tidelift pays maintainers income based on a number of factors, including subscriber usage and strategic importance of a package to the overall health and resilience of the open source software supply chain.

A quick glance at the pages for the mentioned packages shows that these are all 'lifted', in other words "Tidelift is already paying the maintainers of this package":

- https://tidelift.com/lifter/search/npm/tape
- https://tidelift.com/lifter/search/npm/deep-equal
- https://tidelift.com/lifter/search/npm/mock-property
- https://tidelift.com/lifter/search/npm/object-inspect
- https://tidelift.com/lifter/search/npm/object.values

In some cases you are the sole maintainer. In all cases, these packages have dependency graphs (or dev dependency graph, in the `object-inspect` case) including other 'lifted' packages you maintain. As such a person could reasonably infer that you are paid an amount correlated with downloads of these packages.

There's nothing wrong with this in and of itself, of course — I'm very much in favour of people being paid for active, ongoing open source maintenance — but it would seem like a relevant thing to disclose.

**Comment by ljharb** [comment]
My usage of Tidelift is not a secret - i also have GitHub sponsorship. But as you’re well aware there isn’t enough money in the OSS ecosystem to make a difference here, and i wouldn’t be working full time if i could afford to not do so based off open source income.

The reason i don’t go out of my way to mention income explicitly is because it invites people to be paranoid and assume that money is influencing their decisions - I’m sure many people have made absurd accusations about vercel and its imagined influence on svelte, for example - and implying that a paltry amount of money is influencing someone’s technical decisions only serves to further discourage financial contributions to an already criminally underfunded ecosystem.

If there was money in paying an open source maintainer i don’t think I’d have been laid off for 18 months before finding my recent employment.

**Comment by valadaptive** [comment]
Also of note re: Tidelift is that Jordan believes that [packages should publish their test suites, so downstream consumers can run them](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/203), and also that [Tidelift pays per line](https://support.tidelift.com/hc/en-us/articles/4406294816916-How-we-pay-lifters):

> A package containing a large framework gets a larger share than one containing a 2-line function. Currently, the weight is based primarily on code size with some adjustments.

I am not sure if Tidelift includes test code in their package size heuristic, but it is very interesting to note.

**Comment by ljharb** [comment]
I’ve never been aware of Tidelift paying per line that way, and if so I’ll advocate they don’t do that - but also my opinion about shipping tests long predates every funding platform, and is shared by [many other maintainers](https://twitter.com/matteocollina/status/1727651090340954410).

**Comment by ljharb** [comment]
@valadaptive but, that very inflammatory comment is exactly why this isn’t a productive discussion to have.

Assigning motive to anyone receiving money is conspiracy theory territory - I’m not Big Package trying to sell you more Packages

**Comment by valadaptive** [comment]
I think that refusing to elaborate on your motivation for taking over maintainership, state how many other packages you have silently gained commit rights to, or explain why you are volunteering (unpaid) for more and more work that people are literally asking you *not* to do will always invite some form of speculation.

**Comment by HugeLetters** [comment]
@ljharb the problem is that there is a clear mood from the community regarding these changes.
Nobody of active users raised concerns which you've voiced. Majority of people participating in these discussion disagree with you.

Yet instead of saying "Ok, I see that community disagrees with me" and concede you keep pushing against opposing arguments. Even in the other thread where deep-equal was first removed you insisted on some middle-ground which would result in using your package instead of the one suggested by @benmccann 

There was already a discussion with @jessebeach where she concluded that dequal was sufficient for this project - yet now a year later you introduce yours deep-equal-json package even though it was rejected before? 

None of that is a definite proof of anything but hope at least that demonstrates why some people share sentiment that your motivation here is not just a11y concerns.

**Comment by valadaptive** [comment]
> Yet instead of saying "Ok, I see that community disagrees with me" and concede you keep pushing against opposing arguments.

And this isn't even a matter of "well, it's an open-source project, so let the person doing the actual work decide how they want to run the project". Up to this point, all of the code in this project was written by @jessebeach, and people depending on this project assumed that she was the maintainer. You gained maintainership over this project having not made any prior contributions, and the literal first thing you do is add several of your own dependencies to it.

**Comment by Rich-Harris** [comment]
> I’m not Big Package trying to sell you more Packages

Well, that's just the thing — whatever the motivation, in many people's eyes you very much _are_ Big Package trying to sell Packages.

I have no idea whether the sums involved are paltry or not — https://tidelift.com/lifter/search/npm/svelte suggests an estimate of $75 for a package with a little over a million downloads a week, so I would naively expect a package like `get-intrinsic` with 50 million downloads a week to be at least comparable. You maintain a lot of interdependent packages with high download numbers, and you open PRs that switch to your packages over the concerns of users. Again, I'm not making accusations here but I'm sure you can see why people would welcome more transparency.

**Comment by sam-goodwin** [comment]
It seems easier to just copy in the source of dequal and adapt it for node 0.4 than to bring in 16 dependencies just to check if an object is equal, wtf? This is so suss. This has security risk written all over it. 

https://github.com/lukeed/dequal/blob/master/src/index.js

**Comment by ljharb** [comment]
A package of mine, eslint-plugin-jsx-a11y, is the largest single user of this package, by a massive margin.

I could fork the package and then this one would become largely unused, but that’s worse for the ecosystem than what I’m accused of doing here, especially around a11y correctness.

@Rich-Harris your assumption is incorrect; my largest package, qs - which has nearly 72m _weekly_ downloads - 3x more than react, even - pays $50/month, which doesn’t even pay for the cellphone bill that allows me to respond from a grocery store on a weekend. Nobody is getting rich off of open source and I’m astonished you’re not thoroughly aware of this fact.

**Comment by valadaptive** [comment]
OK, a package with 72 million downloads/month gets $50/month. That's $0.69 per month per million downloads per week (forgive me if I've fumbled my units). Looking at [your NPM stats](https://my-npm-stats.org/?author=ljharb), your packages, all combined, net 1736 million downloads a week. That's $1200/month, or $14,000 a year. A far cry from a living wage to be sure, but significant supplemental income.

**Comment by Rich-Harris** [comment]
You're pretending I said something I didn't.

**Comment by ljharb** [comment]
@Rich-Harris youre being very careful to not directly accuse of me anything - but bringing money into the discussion IS an accusation. It’s simply irrelevant. I’d be doing the exact same things even if it cost me money to do them - my motives are pure.

@valadaptive i live in the SF Bay Area, and i have children. It is in fact insignificant supplemental income; if i was grubbing for money I’d spend the time contracting and make 10+ times that.

**Comment by pngwn** [comment]
@ljharb as a long time supporter of your work and everything you have done for the JavaScript ecosystem, I have to say I am a little disappointed in your defensive response to calls for greater transparency. 

I do not know what your motivations are here but surely we can all agree that more transparency is in the best interests of the broader ecosystem. Additionally, I’m a little disappointed that the opinions of so many folks from the community are being disregarded here and you are taking a stance of “what I say goes” and refusing to entertain other opinions at all. 

Like I say, I don’t know what your motivations are, nor do I need to, because it is clear that your actions and behaviours are causing active harm to the JavaScript and open source ecosystem at this point. 

Very disappointing.

**Comment by mikkurogue** [comment]
I still don't fully understand or see the value of supporting an outdated node version with a new update. Surely whoever is still their app on that node version isn't just updating packages? 

To me this all seems kind of unnecessary and frankly bit of a waste of time not only for the author but also for everyone involved having to argue about what is best for "ancient" engines who have long since lost LTS status. Isn't that the whole idea of LTS? Or am I just plainly misunderstanding the problem at its core?

**Comment by ANorseDude** [comment]
It's obvious that @ljharb got caught red handed with his double play. Please stop;, lock the thread, and move one.

**Comment by SoaresMG** [comment]
Those dependencies that it was changed to (e.g deep-equal-json) are they more performant as the previous deps?

IMO that's more important than supporting a node version with 10 years.

**Comment by SukkaW** [comment]
That's why I started [nolyfill](https://github.com/SukkaW/nolyfill) exactly a year ago.

nollyfill empowers npm overrides / yarn resolutions / pnpm overrides to replace all ljharb's chunky and slow polyfills with native implementation. nolyfill can speed up the package installation process, reduce disk usage, and extend the lifespan of precious SSDs.

**Comment by remie** [comment]
@ljharb while I fully appreciate having strong opinions, I do think it's pretty callous to disregard other maintainers opinions, especially with regard to a topic on which they already made their stance clear (see https://github.com/A11yance/aria-query/pull/497).

I think in this case you really should have tried to get consensus from other maintainers before merging, and specifically that from @jessebeach.

Not doing so makes this change look pedantic at best and malicious at worst, neither are good qualities for a maintainer. This is not proper stewardship.

**Comment by ImLunaHey** [comment]
someone brought up a really important issue here https://x.com/samgoodwin89/status/1804577766232437005

Screenshot 2024-06-23 at 7 23 35 PM

**Comment by gm112** [comment]
This seems really suspicious that you would merge many dependencies that you have a financial incentive, with no real good technical reason. I hope Svelte removes this dependency either by forking or doing something. Forking isn't ideal, but its better than dealing with an uncooperative maintainer who places personal needs above others. Also, the gaslighting isn't a good look, either.

**Comment by valadaptive** [comment]
> someone brought up a really important issue here https://x.com/samgoodwin89/status/1804577766232437005
> Screenshot 2024-06-23 at 7 23 35 PM

I looked at this briefly, and right now I don't think there's any way to inject a backdoor via compromising the GitHub action. AFAIK, there are no NPM credentials stored in this repo and no way to publish via CI--it's only used for running tests.

**Comment by Malix-off** [comment]
At first, I didn't really mind about that PR (and quite frankly didn't know how Svelte used it precisely)

But seeing valid [legitimity](https://github.com/A11yance/axobject-query/pull/354#issuecomment-2183910073), [supply chain threat](https://github.com/A11yance/axobject-query/pull/354#issuecomment-2184042343) and [conflicts of interests](https://github.com/A11yance/axobject-query/pull/354#issuecomment-2184109808) concerns; also adding that [this PR is the first ever contribution to this project from them](https://github.com/A11yance/axobject-query/commits/3.x?author=ljharb);

And more importantly seeing how the Author replied _(which I personally consider borderline gaslighting imho)_ coupled to no questioning on his part considering there is 157 downvotes currently while nobody backing him

Made me very skeptical about Author's intentions and concerned about how dependants (i.e. Svelte) and transitive dependants (i.e. users of Svelte) will handle the situation

**Comment by mikkurogue** [comment]
> > someone brought up a really important issue here https://x.com/samgoodwin89/status/1804577766232437005
> > Screenshot 2024-06-23 at 7 23 35 PM
> 
> I looked at this briefly, and right now I don't think there's any way to inject a backdoor via compromising the GitHub action. AFAIK, there are no NPM credentials stored in this repo and no way to publish via CI--it's only used for running tests.

I do think this does warrant to be at least a cautionary view and probably requires more in-depth on the reasoning for author to want to interject his own packages into something when the issue was first relatively trivial making this come across as just malicious compliance at best and downright petty.

File: .github/workflows/test.yml
Comment (gm112) [reply .github/workflows/test.yml:25]: Easy way to sneak a backdoor in lol

**Comment by SukkaW** [comment]
> I hope Svelte removes this dependency either by forking or doing something. Forking isn't ideal, but it's better than dealing with an uncooperative maintainer who places personal needs above others.

[As I mentioned earlier](https://github.com/A11yance/axobject-query/pull/354#issuecomment-2184922910), I started [nolyfill](https://github.com/SukkaW/nolyfill) a year go for the exact same reason:

> nollyfill empowers npm overrides / yarn resolutions / pnpm overrides to replace all ljharb's chunky and slow polyfills with lightweight packages that are alias to native built-in implementation. nolyfill can speed up the package installation process, reduce disk usage, and extend the lifespan of precious SSDs.

But I am also a single maintainer behind the project and I am not a full-time open-source contributor/maintainer currently. If anyone wants to join and renovate the nolyfill project, feel free to contact me and together we can make nolyfill a strong force aginst this absurd.

**Comment by pngwn** [comment]
Actions should never track main unless it is single use action specific to one repo. This is a terrible security practice and should never be done anywhere.

**Comment by ImLunaHey** [comment]
> Actions should never track main unless it is single use action specific to one repo. This is a terrible security practice and should never be done anywhere.

which is why im not so sure this wasnt intentional.. they should know better.

**Comment by valadaptive** [comment]
> But I am also a single maintainer behind the project and I am not a full-time open-source contributor/maintainer currently. If anyone wants to join and renovate the nolyfill project, feel free to contact me and together we can make nolyfill a strong force aginst this absurd.

@SukkaW, maybe get in touch with @43081j, who is currently working on lighter alternatives to many npm packages, and has created [an eslint plugin which suggests replacements for dependencies](https://github.com/es-tooling/eslint-plugin-depend) and [an ad-hoc effort to clean up the JS ecosystem](https://github.com/43081j/ecosystem-cleanup).

**Comment by alper** [comment]
How is this merged?

**Comment by mikkurogue** [comment]
> > Actions should never track main unless it is single use action specific to one repo. This is a terrible security practice and should never be done anywhere.
> 
> which is why im not so sure this wasnt intentional.. they should know better.

especially if you read his github intro thing, you'd think he'd know better then...

**Comment by pngwn** [comment]
Indeed. Security is an ‘ethical’ obligation in my view.

**Comment by tigerabrodi** [comment]
😬

File: package.json
Comment (gtm-nayan) [inline package.json:67]: Why does it list engines as `node >= 0.4` but only test `>= 0.8` in the tests? Seems counter-intuitive to the spirit of the PR.

File: .github/workflows/test.yml
Comment (appleseedexm) [inline .github/workflows/test.yml:17]: Building test-only libs into `lib` seems like bad and potentially dangerous design to me. Better to use something like `lib-test`. Dont want future changes to lead to chained attacks like runtime depending on `lib` and tests modifying them.

**Comment by ashley0143** [comment]
why was this merged?? lmao

**Comment by sathwik77** [comment]
The fact that this PR got merged with 194 down votes is just hilarious 😂

**Comment by valadaptive** [comment]
I can appreciate how awestruck everyone is, but I can't help but worry that "omg who did this??? 😂😂😂" type comments will cause this discussion to be locked, which is a shame since there is a lot of real discussion about the JS ecosystem going on here.

**Comment by ashley0143** [comment]
@valadaptive true, im sorry :p

**Pull request merged**
